src/Routers/UsersRouter.js

```diff
@@ -453,6 +479,97 @@ export class UsersRouter extends ClassesRouter {
     });
   }
 
+  async handleChallenge(req) {
+    const { username, email, password, authData, challengeData } = req.body;
+
+    // if username or email provided with password try to find the user with default
+    // system
+    let user;
+    if (username || email) {
+      if (!password)
+        throw new Parse.Error(
+          Parse.Error.OTHER_CAUSE,
+          'You provided username or email, you need to also provide password.'
+        );
+      user = await this._authenticateUserFromRequest(req);
+    }
+
+    if (!challengeData) throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'Nothing to challenge.');
+
+    if (typeof challengeData !== 'object')
+      throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'challengeData should be an object.');
+
+    // Try to find user by authData
+    if (authData) {
+      if (typeof authData !== 'object')
+        throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'authData should be an object.');
+      // To avoid security issue we should only support one identifying method
+      if (user)
+        throw new Parse.Error(
+          Parse.Error.OTHER_CAUSE,
+          'You cant provide username/email and authData, only use one identification method.'
+        );
+
+      if (Object.keys(authData).filter(key => authData[key].id).length > 1) {
+        throw new Parse.Error(
+          Parse.Error.OTHER_CAUSE,
+          'You cant provide more than one authData provider with an id.'
+        );
+      }
+
+      const results = await Auth.findUsersWithAuthData(req.config, authData);
+
+      try {
+        if (!results[0] || results.length > 1)
+          throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'User not found.');
+
+        // Find the provider used to find the user
+        const provider = Object.keys(authData).find(key => authData[key].id);
+
+        // Validate authData used to identify the user
+        // to avoid guess id attack
+        const { validator } = req.config.authDataManager.getValidatorForProvider(provider);
+        await validator(
+          authData[provider],
+          { config: req.config, auth: req.auth, isChallenge: true },
+          Parse.User.fromJSON({ className: '_User', ...results[0] })
+        );
+        user = results[0];
+      } catch (e) {
+        // Rewrite the error to avoid guess id attack
+        logger.error(e);
+        throw new Parse.Error(Parse.Error.OTHER_CAUSE, 'User not found.');
+      }
+    }
+
+    // Execute challenge step by step
```

Comment:
// Execute challenge with consistent order

Can you briefly explain in the comment why this is relevant?